src/analyze/analyze-unit-gdb-start.c

```diff
@@ -0,0 +1,299 @@
+                const char *name,
+                const char *mode,
+                sd_bus_error *error,
+                BusWaitForJobs *w) {
+
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        const char *path;
+        int r;
+
+        assert(method);
+        assert(name);
+        assert(mode);
+        assert(error);
+
+        log_debug("%s dbus call org.freedesktop.systemd1.Manager %s(%s, %s)",
+                  "Executing", method, name, mode);
+
+        r = bus_call_method(bus, bus_systemd_mgr, method, error, &reply, "ss", name, mode);
+        if (r < 0)
+                return log_error_errno(r, "Failed to start %s: %s", name, bus_error_message(error, r));
+
+        r = sd_bus_message_read(reply, "o", &path);
+        if (r < 0)
+                return bus_log_parse_error(r);
+        if (w) {
+                log_debug("Adding %s to the set", path);
+                r = bus_wait_for_jobs_add(w, path);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to watch job for %s: %m", name);
+        }
+
+        r = bus_wait_for_jobs_one(w, path, /* flags= */ BUS_WAIT_JOBS_LOG_ERROR, /* extra_args= */ NULL);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+int verb_unit_gdb_start(int argc, char *argv[], void *userdata) {
+        static const struct sigaction sa = {
+                .sa_sigaction = sigterm_process_group_handler,
+                .sa_flags = SA_SIGINFO,
+        };
+
+        _cleanup_free_ char *unit = NULL;
+        int r;
+
+        if (arg_transport != BUS_TRANSPORT_LOCAL)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Cannot spawn a debugger for a remote service");
+
+        r = unit_name_mangle_with_suffix(argv[1], "as unit", UNIT_NAME_MANGLE_WARN, ".service", &unit);
+        if (r < 0)
+                return log_error_errno(r, "Cannot mangle unit name: %m");
+
+        if (!arg_debugger) {
+                arg_debugger = strdup(secure_getenv("SYSTEMD_DEBUGGER") ?: "gdb");
+                if (!arg_debugger)
+                        return log_oom();
+        }
+
+        if (!STR_IN_SET(arg_debugger, "gdb", "lldb"))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "The debugger must be either 'gdb' or 'lldb'");
+
+        _cleanup_strv_free_ char **debugger_call = NULL;
+        r = strv_extend(&debugger_call, arg_debugger);
+        if (r < 0)
+                return log_oom();
+
+        /* Don't interfere with debugger and its handling of SIGINT. */
+        (void) ignore_signals(SIGINT);
+        (void) sigaction(SIGTERM, &sa, NULL);
+
+        _cleanup_free_ char *fork_name = strjoin("(", debugger_call[0], ")");
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        r = pidref_safe_fork(fork_name, FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL, &pidref);
+        if (r < 0)
+                return log_error_errno(r, "Fork failed: %m");
+
+        if (r == 0) {
```

Comment:
so, fork() is an ugly beast, because it copies the process in its entire state, regardless what that might mean, and then is a CoW-trap. Hence, fork() is ok to use, but the child should be as short lived as we can get away with, until it calls execve(). or in other words: please make sure to move all the dbus stuff below into the parent, do it before you fork, and then only fork once you have everything ready.